File: Course4/Android/Lab3/app/src/main/java/com/zinchenko/lab3/Calculator.java
package com.zinchenko.lab3;

import android.util.Log;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.StringTokenizer;

public final class Calculator
{
	private final String operators = "+-*/^";
	private final String delimiters = "() " + operators;

	public final String process(String input) throws RuntimeException
	{
		Log.i("Test", input + " -> " + parse(input).toString() + " = " + calculate(parse(input)));
		return String.valueOf(calculate(parse(input)));
	}

	private List<String> parse(String infix) throws RuntimeException
	{
		List<String> postfix = new ArrayList<>();
		Deque<String> stack = new ArrayDeque<>();
		StringTokenizer tokenizer = new StringTokenizer(infix, delimiters, true);
		String prev = "";
		String curr = "";

		while (tokenizer.hasMoreTokens())
		{
			curr = tokenizer.nextToken();
			if (!tokenizer.hasMoreTokens() && isOperator(curr))
				throw new RuntimeException("Некорректное выражение: " + infix);

			if (curr.equals(" "))
				continue;

			if (isDelimiter(curr))
			{
				if (curr.equals("("))
					stack.push(curr);
				else if (curr.equals(")"))
				{
					while (!stack.peek().equals("("))
					{
						postfix.add(stack.pop());
						if (stack.isEmpty())
							throw new RuntimeException("Скобки не согласованы: " + infix);
					}
					stack.pop();
				}
				else
				{
					if (curr.equals("-") && (prev.equals("") || (isDelimiter(prev) && !prev.equals(")"))))
						curr = "u-";
					else
						while (!stack.isEmpty() && (getPriority(curr) <= getPriority(stack.peek())))
							postfix.add(stack.pop());
					stack.push(curr);
				}
			}
			else
				postfix.add(curr);
			prev = curr;
		}

		while (!stack.isEmpty())
			if (isOperator(stack.peek()))
				postfix.add(stack.pop());
			else
				throw new RuntimeException("Скобки не согласованы: " + infix);

		return postfix;
	}

	private double calculate(List<String> postfix)
	{
		Deque<Double> stack = new ArrayDeque<>();
		Double a;
		Double b;

		for (String str : postfix)
		{
			switch (str)
			{
				case "+":
					b = stack.pop();
					a = stack.pop();
					stack.push(a + b);
					break;
				case "-":
					b = stack.pop();
					a = stack.pop();
					stack.push(a - b);
					break;
				case "*":
					b = stack.pop();
					a = stack.pop();
					stack.push(a * b);
					break;
				case "/":
					b = stack.pop();
					a = stack.pop();
					stack.push(a / b);
					break;
				case "^":
					b = stack.pop();
					a = stack.pop();
					stack.push(Math.pow(a, b));
					break;
				case "u-":
					a = stack.pop();
					stack.push(-a);
					break;
				default:
					stack.push(Double.valueOf(str));
			}
		}
		return stack.pop();
	}

	private boolean isDelimiter(String token)
	{
		if (token.length() != 1)
			return false;
		else
			return delimiters.indexOf(token.charAt(0)) != -1;
	}

	private boolean isOperator(String token)
	{
		if (token.equals("u-"))
			return true;
		else
			return operators.indexOf(token.charAt(0)) != -1;
	}

	private byte getPriority(String token)
	{
		switch (token.charAt(0))
		{
			case '(':
				return 1;
			case '+':
			case '-':
				return 2;
			case '*':
			case '/':
				return 3;
			case '^':
				return 4;
			default:
				return 5;
		}
	}
}

File: Course4/AT/Lab2/app/src/main/java/com/zinchenko/lab2/Calculator.kt
package com.zinchenko.lab2

import java.lang.RuntimeException
import java.util.*
import kotlin.Throws
import kotlin.math.pow

class Calculator
{
	private val operators = "+-*/^"
	private val delimiters = "() $operators"

	@Throws(RuntimeException::class)
	fun process(input: String): String
	{
		return calculate(parse(input)).toString()
	}

	@Throws(RuntimeException::class)
	private fun parse(infix: String): List<String>
	{
		val postfix = ArrayList<String>()
		val stack = ArrayDeque<String>()
		val tokenizer = StringTokenizer(infix, delimiters, true)
		var prev = ""
		var curr: String

		while (tokenizer.hasMoreTokens())
		{
			curr = tokenizer.nextToken()
			if (!tokenizer.hasMoreTokens() && isOperator(curr)) throw RuntimeException("Некорректное выражение: $infix")
			if (curr == " ") continue
			if (isDelimiter(curr))
			{
				when (curr)
				{
					"(" -> stack.push(curr)
					")" ->
					{
						while (stack.peek() != "(")
						{
							postfix.add(stack.pop())
							if (stack.isEmpty()) throw RuntimeException("Скобки не согласованы: $infix")
						}
						stack.pop()
					}
					else ->
					{
						if (curr == "-" && (prev == "" || isDelimiter(prev) && prev != ")")) curr = "u-"
						else while (!stack.isEmpty() && getPriority(curr) <= getPriority(stack.last())) postfix.add(
							stack.pop()
						)
						stack.push(curr)
					}
				}
			}
			else postfix.add(curr)
			prev = curr
		}
		while (!stack.isEmpty()) if (isOperator(stack.last())) postfix.add(stack.pop())
		else throw RuntimeException(
			"Скобки не согласованы: $infix"
		)
		return postfix
	}

	private fun calculate(postfix: List<String>): Double
	{
		val stack: Deque<Double> = ArrayDeque()
		var a: Double
		var b: Double
		for (str in postfix) when (str)
		{
			"+" ->
			{
				b = stack.pop()
				a = stack.pop()
				stack.push(a + b)
			}
			"-" ->
			{
				b = stack.pop()
				a = stack.pop()
				stack.push(a - b)
			}
			"*" ->
			{
				b = stack.pop()
				a = stack.pop()
				stack.push(a * b)
			}
			"/" ->
			{
				b = stack.pop()
				a = stack.pop()
				stack.push(a / b)
			}
			"^" ->
			{
				b = stack.pop()
				a = stack.pop()
				stack.push(a.pow(b))
			}
			"u-" ->
			{
				a = stack.pop()
				stack.push(-a)
			}
			else -> stack.push(java.lang.Double.valueOf(str))
		}
		return stack.pop()
	}

	private fun isDelimiter(token: String): Boolean
	{
		return if (token.length != 1) false else delimiters.indexOf(token[0]) != -1
	}

	private fun isOperator(token: String): Boolean
	{
		return if (token == "u-") true else operators.indexOf(token[0]) != -1
	}

	private fun getPriority(token: String): Byte
	{
		return when (token[0])
		{
			'(' -> 1
			'+', '-' -> 2
			'*', '/' -> 3
			'^' -> 4
			else -> 5
		}
	}
}